crates/bevy_ecs/src/system/commands/entity_command.rs

```diff
@@ -74,96 +76,75 @@ use bevy_ptr::OwningPtr;
-    /// footprint than `(Entity, Self)`.
-    /// In most cases the provided implementation is sufficient.
-    #[must_use = "commands do nothing unless applied to a `World`"]
-    fn with_entity(self, entity: Entity) -> impl Command<(Result, CommandError)>
-    where
-        Self: Sized,
+impl<C: EntityCommand> CommandWithEntity<Result<(), EntityFetchError>> for C {
+    fn with_entity(
+        self,
+        entity: Entity,
+    ) -> impl Command<Result<(), EntityFetchError>> + HandleError<Result<(), EntityFetchError>>
     {
-        move |world: &mut World| -> Result<(), CommandError> {
-            if world.entities().contains(entity) {
-                match self.apply(entity, world) {
-                    Ok(_) => Ok(()),
-                    Err(error) => Err(CommandError::CommandFailed(error)),
-                }
-            } else {
-                Err(CommandError::NoSuchEntity(
-                    entity,
-                    world
-                        .entities()
-                        .entity_does_not_exist_error_details_message(entity),
-                ))
-            }
+        move |world: &mut World| -> Result<(), EntityFetchError> {
+            let entity = world.get_entity_mut(entity)?;
+            self.apply(entity);
+            Ok(())
         }
     }
 }
 
-impl<F> EntityCommand for F
-where
-    F: FnOnce(Entity, &mut World) + Send + 'static,
-{
-    fn apply(self, id: Entity, world: &mut World) -> Result {
-        self(id, world);
-        Ok(())
-    }
-}
-
-impl<F> EntityCommand<Result> for F
-where
-    F: FnOnce(Entity, &mut World) -> Result + Send + 'static,
+impl<
+        C: EntityCommand<Result<T, Err>>,
+        T,
+        Err: core::fmt::Debug + core::fmt::Display + Send + Sync + 'static,
+    > CommandWithEntity<Result<T, EntityCommandError<Err>>> for C
 {
-    fn apply(self, id: Entity, world: &mut World) -> Result {
-        self(id, world)
+    fn with_entity(
+        self,
+        entity: Entity,
+    ) -> impl Command<Result<T, EntityCommandError<Err>>> + HandleError<Result<T, EntityCommandError<Err>>>
+    {
+        move |world: &mut World| {
+            let entity = world.get_entity_mut(entity)?;
+            self.apply(entity).map_err(EntityCommandError::Error)
+        }
     }
 }
 
-impl<F> EntityCommand<World> for F
-where
-    F: FnOnce(EntityWorldMut) + Send + 'static,
-{
-    fn apply(self, id: Entity, world: &mut World) -> Result {
-        self(world.entity_mut(id));
-        Ok(())
-    }
+/// An error that occurs when running an [`EntityCommand`] on a specific entity.
+#[derive(thiserror::Error, Debug)]
+pub enum EntityCommandError<E> {
+    /// The entity this [`EntityCommand`] tried to run on could not be fetched.
+    #[error(transparent)]
+    EntityFetchError(#[from] EntityFetchError),
+    /// An error that occurred while running the [`EntityCommand`].
+    #[error("{0}")]
+    Error(E),
```

Comment:
This variant name is pretty bad. `EntityCommandError::Error`? I liked the `CommandFailed` name from the last PR.